File: delivery/api/src/main/java/org/example/delivery/api/common/exception/ApiException.java
package org.example.delivery.api.common.exception;

import lombok.Getter;
import org.example.delivery.api.common.error.ErrorCode;

@Getter
public class ApiException extends RuntimeException implements CustomException {

    private final ErrorCode errorCode;
    private final String errorDescription;

    public ApiException(ErrorCode errorCode) {
        super(errorCode.getDescription());
        this.errorCode = errorCode;
        this.errorDescription = errorCode.getDescription();
    }

    public ApiException(ErrorCode errorCode, String errorDescription) {
        super(errorCode.getDescription());
        this.errorCode = errorCode;
        this.errorDescription = errorDescription;
    }

    public ApiException(ErrorCode errorCode, Throwable tx) {
        super(tx);
        this.errorCode = errorCode;
        this.errorDescription = errorCode.getDescription();
    }

    public ApiException(ErrorCode errorCode, Throwable tx, String errorDescription) {
        super(tx);
        this.errorCode = errorCode;
        this.errorDescription = errorDescription;
    }
}

File: delivery/common/src/main/kotlin/org/example/delivery/common/exception/ApiException.kt
package org.example.delivery.common.exception

import org.example.delivery.common.error.ErrorCode
import java.lang.RuntimeException

class ApiException : RuntimeException, CustomException {

    override val errorCode: ErrorCode?
    override val errorDescription: String?

    constructor(errorCode: ErrorCode): super(errorCode.getDescription()) {
        this.errorCode = errorCode
        this.errorDescription = errorCode.getDescription()
    }

    constructor(
        errorCode: ErrorCode,
        errorDescription: String,
    ): super(errorCode.getDescription()) {
        this.errorCode = errorCode
        this.errorDescription = errorDescription
    }

    constructor(
        errorCode: ErrorCode,
        tx: Throwable,
    ): super(tx) {
        this.errorCode = errorCode
        this.errorDescription = errorCode.getDescription()
    }

    constructor(
        errorCode: ErrorCode,
        tx: Throwable,
        errorDescription: String,
    ): super(tx) {
        this.errorCode = errorCode
        this.errorDescription = errorDescription
    }
}